src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -131,7 +131,20 @@ def _bigquery(self) -> bigquery.Client:
             return bigquery.Client(project=self._source.project)
 
     def _run_sql(self, query: str) -> BigQueryRows:
-        return self._bigquery.query(query)
+        delays = (10, 20, 40, 80)
+        assert sum(delays) < config.contribution_lambda_timeout
+        for attempt, delay in enumerate((*delays, None)):
+            job = self._bigquery.query(query)
+            try:
+                return job.result()
+            except Forbidden as e:
+                if 'Exceeded rate limits' in e.message and delay is not None:
+                    log.warning('Exceeded BigQuery rate limit during attempt %i/%i. Retrying in %is.',
+                                attempt + 1, len(delays), delay, exc_info=e)
+                    time.sleep(delay)
+                else:
+                    raise e
+        assert False
```

Comment:
It seems that this `assert False` here just to ensure we never exit without either: 
1. returning a query_job with a valid result
2. raise an exception/error back to the caller 

But even in the case where we set the delays to `()` and there is no job.result(),  ln 146 ` raise e ` should  still be hit, is this an intentional extra precaution?

Reply:
Yes. The `assert False` should never be hit. That's exactly why it is there. Assertions are used to communicate to readers of the code which invariants the developer intends to establish. With this assert I am communicating to you that the loop never exits naturally.

Reply:
Yes. The `assert False` should never be hit. That's exactly why it is there. Assertions are used to communicate to readers of the code which invariants the developer intends to establish. With this assert I am communicating to you that the loop never exits naturally.